src/Microsoft.DotNet.MSBuildSdkResolver/FXVersion.cs

```diff
@@ -40,18 +39,101 @@ public static int Compare(FXVersion s1, FXVersion s2)
+                // Empty (release) is higher precedence than prerelease
+                return string.IsNullOrEmpty(s1.Pre) ? (string.IsNullOrEmpty(s2.Pre) ? 0 : 1) : -1;
             }
 
-            int preCompare = string.CompareOrdinal(s1.Pre, s2.Pre);
-            if (preCompare != 0)
+            string [] ids1 = s1.Pre.Substring(1).Split('.');
+            string [] ids2 = s2.Pre.Substring(1).Split('.');
+
+            for (int i = 0; true ; ++i)
+            {
+                if ((i >= ids1.Length) || (i >= ids2.Length))
+                {
+                    // One or both of the identifier lists has terminated
+                    // Longest list of identifiers has highest precedecnce
+                    return ids1.Length == ids2.Length ? 0 : ids1.Length >= ids2.Length ? 1 : -1;
+                }
+
+                int preCompare = string.CompareOrdinal(ids1[i], ids2[i]);
+
+                if (preCompare != 0)
+                {
+                    int n1;
+                    bool b1 = int.TryParse(ids1[i], out n1);
+                    int n2;
+                    bool b2 = int.TryParse(ids2[i], out n2);
+
+                    if (b1 && b2)
+                    {
+                        // Both are numeric use numeric ordering (note never equal)
+                        return n1 > n2 ? 1 : -1;
+                    }
+                    else if (b1 || b2)
+                    {
+                        // Exactly one is numeric
+                        // Numeric has lower precedence
+                        return b2 ? 1 : -1;
+                    }
+                    else
+                    {
+                        // Both are alphanumeric, use ascii sort order
+                        // Since we are using only ascii characters, unicode ordinal sort == ascii sort
+                        return preCompare > 0 ? 1 : -1;
+                    }
+                }
+            }
+        }
+
+        private static bool ValidIdentifier(string id, bool buildMeta)
+        {
+            if (string.IsNullOrEmpty(id))
+            {
+                // Identifier must not be empty
+                return false;
+            }
+
+            if (id.FindFirstNotOf("-0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ", 0) != -1)
+            {
+                // ids must be of the set [0-9a-zA-Z.-]
+                return false;
+            }
+
+            if (!buildMeta && id.Substring(0,1) == "0" && id.Length > 1 && id.FindFirstNotOf("0123456789", 1) == -1)
+            {
+                // numeric identifiers must not be padded with 0s
+                return false;
+            }
+            return true;
+        }
+
+        private static bool ValidIdentifiers(string ids)
+        {
+            if (string.IsNullOrEmpty(ids))
+            {
+                return true;
+            }
+
+            bool prerelease = ids.Substring(0, 1) == "-";
```

Comment:
This code allocates temporary strings on the heap left, right, and center.

I want to keep it as close to core-setup as possible, but the common case of creating strings to check first char is very low hanging fruit.

On, .NET Core, this can be ids.StartsWith('-') (note, char not string arg).

This has to target netstandard2.0/net4.7.2, though, we can polyfill it in StringExtensions since it's trivially `(s.Length > 0 && s[0] == c)`

Reply:
Actually, I think everywhere you use `x.Substring(0, 1)` == "y"` can be `x[0] == 'y'` since `Substring(0, 1)` throws on empty.

Reply:
Looks like the native code is already like this. Please keep it as close as possible.